docs/images.md

```diff
@@ -5,31 +5,157 @@ You create your Docker image and push it to a registry before referring to it in
-Kubernetes has native support for the [Google Container Registry](https://cloud.google.com/tools/container-registry/), when running on Google Compute Engine.  If you are running your cluster on Google Compute Engine or Google Container Engine, simply use the full image name (e.g. gcr.io/my_project/image:tag) and the kubelet will automatically authenticate and pull down your private image.
+
+Kubernetes has native support for the [Google Container
+Registry](https://cloud.google.com/tools/container-registry/), when running on Google Compute
+Engine (GCE).  If you are running your cluster on GCE or Google Container Engine (GKE), simply
+use the full image name (e.g. gcr.io/my_project/image:tag) and the kubelet will automatically
+authenticate and pull down your private image.
+
+All pods in a cluster will have read access to images in this registry.
 
 ### Other Private Registries
-Docker stores keys for private registries in a `.dockercfg` file.  Create a config file by running `docker login <registry>.<domain>` and then copying the resulting `.dockercfg` file to the kubelet working dir.
-The kubelet working dir varies by cloud provider.  It is `/` on GCE and `/home/core` on CoreOS.  You can determine the working dir by running this command:
-`sudo ls -ld /proc/$(pidof kubelet)/cwd` on a kNode.
+Docker stores keys for private registries in a `.dockercfg` file.  Create a config file by running
+`docker login <registry>.<domain>` and then copy the resulting `.dockercfg` file to the kubelet
+working dir on each node in your cluster.  The kubelet working dir varies by cloud provider.  It is `/` on GCE and `/home/core`
+on CoreOS.  You can determine the working dir by running this command: `sudo ls -ld /proc/$(pidof
+kubelet)/cwd` on a node.
+
+You must ensure all nodes in the cluster have the same `.dockercfg`.  Otherwise, pods will run on
+some nodes and fail to run on others.  For example, if you use node autoscaling, then each instance
+template needs to include the `.dockercfg` or mount a drive that contains it.
 
-All users of the cluster will have access to any private registry in the `.dockercfg`.
+All pods will have read access to images in any private registry with keys in the `.dockercfg`.
 
-## Preloading Images
+### Pre-pulled Images
 
 Be default, the kubelet will try to pull each image from the specified registry.
 However, if the `imagePullPolicy` property of the container is set to `IfNotPresent` or `Never`,
 then a local image is used (preferentially or exclusively, respectively).
 
+If you want to rely on pre-pulled images as a substitute for registry authentication,
+you must ensure all nodes in the cluster have the same pre-pulled images.
+
 This can be used to preload certain images for speed or as an alternative to authenticating to a private registry.
 
-Pull Policy is per-container, but any user of the cluster will have access to all local images.
+All pods will have read access to any pre-pulled images.
```

Comment:
----- Original Message -----
> >  This can be used to preload certain images for speed or as an alternative
> >  to authenticating to a private registry.
> >  
> > -Pull Policy is per-container, but any user of the cluster will have access
> > to all local images.
> > +All pods will have read access to any pre-pulled images.
> 
> > Are you saying that if a container uses PullIfNotPresent, that it
> > potentially has access to all images ever pulled on the machine, including
> > those pulled with imagePullSecret?
> 
> Yes, this is a side-effect of not having the equivalent of the Docker
> registry not having a SubjectAccessReview (MotherMayI) API.  It will work
> for the same reason that pre-pulling images will work.
> 
> @smarterclayton Suggested a couple workarounds, but they all behaved
> unexpectedly when access to the Docker registry was down, but not denied.
> Things like "try to pull the image and make sure you start getting a layer
> back".

Also, this fits in with something we've been discussing with @ncdc with regards to push - the daemon essentially has to do proof of work to the registry when pushing images (prove that it has the lower level layers) otherwise you'd be able to lie about what images you have access to and push/pull your own.  That proof of work today is "push the whole binary to the server", but there are ongoing discussions about it being cryptographic "I ran a crypto-secure transformation on something locally and it proves I have the binary" or even more simply, passing authorization on push that proves access to the other image "here is the token I used to fetch the image".  

The latter is a lot like the pullIfNotPresent problem - the Kubelet needs to avoid being a confused deputy (I pulled A for Alice, but Bob shouldn't be able to leverage that to get A) by checking Bob's credentials against the registry for A first.

> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/9495/files#r32212336